src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+        std::list<Motion*> leavesToPrune;
+
+        // Get the list from the NN structure
+        nn_->list(treeVertices);
+
+        // Iterate over the list of vertices
+        for (std::vector<Motion*>::iterator vIter = treeVertices.begin(); vIter != treeVertices.end(); ++vIter)
+        {
+            // Check if it passes the pruning condition
+            if (leafPruningCondition(*vIter, pruneTreeCost) == true)
+            {
+                // Add to the list of leaves to prune
+                leavesToPrune.push_back(*vIter);
+            }
+            // No else, does not meet the pruning criteria
+        }
+
+        // Iterate over the list of leaves to prune
+        std::list<Motion*>::iterator lpIter = leavesToPrune.begin();
+
+        while (lpIter != leavesToPrune.end())
         {
-            pruneScratchSpace_.newTree.push_back(candidate);
-            pruneScratchSpace_.candidates.insert(pruneScratchSpace_.candidates.end(),
-                candidate->children.begin(), candidate->children.end());
+            // Remove the leaf from the nn_ structure
+            nn_->remove(*lpIter);
+
+            // Remove the leaf from its parent
+            removeFromParent(*lpIter);
+
+            // Check if its parent should be added to the list
+            if (leafPruningCondition((*lpIter)->parent, pruneTreeCost) == true)
+            {
+                // Add to the list of leaves to prune
+                leavesToPrune.push_back((*lpIter)->parent);
+            }
+            // No else, does not meet the pruning criteria
+
+            // Erase the actual motion, by first freeing the state
+            si_->freeState((*lpIter)->state);
+
+            // then deleting the pointer
+            delete *lpIter;
+
+            // And finally removing it from the list, erase returns the next iterator
+            lpIter = leavesToPrune.erase(lpIter);
+
+            // Update our counter
+            ++numPruned;
+        }
+
+        // Update the cost at which we've pruned:
+        prunedCost_ = pruneTreeCost;
+
+        // And if we're using informed sampling, the measure to which we've pruned
+        if (useInformedSampling_)
+        {
+            prunedInfMeasure_ = infSampler_->getInformedMeasure(prunedCost_);
         }
     }
 
-    // To create the new nn takes one order of magnitude in time more than just checking how many
-    // states would be pruned. Therefore, only prune if it removes a significant amount of states.
-    if ((double)pruneScratchSpace_.newTree.size() / tree_size < pruneStatesThreshold_)
-    {
-        for (std::size_t i = 0; i < pruneScratchSpace_.toBePruned.size(); ++i)
-            deleteBranch(pruneScratchSpace_.toBePruned[i]);
+    return numPruned;
+}
 
-        nn_->clear();
-        nn_->add(pruneScratchSpace_.newTree);
 
-        return (tree_size - pruneScratchSpace_.newTree.size());
+bool ompl::geometric::RRTstar::leafPruningCondition(Motion *motion, const base::Cost& pruneTreeCost)
```

Comment:
The pruning condition in one place for ease of use on both the vertex and the vertex's parent.